Add Crafty.viewport.absoluteScale()

mildmojo opened the pull request:
### Resolution independence: The problem

I need my games to scale to fit the browser window, or at least select something close from a list of known good resolutions. Unfortunately, Crafty's viewport system is working against me. Looking at other open issues, it seems like a common pain point.
### Crafty today

I can't figure out the use case for `Crafty.viewport.scale`. It doesn't work at all for my purposes. Where it applies a cumulative scale factor, I need to apply an absolute scale factor calculated from the window size. Additionally, Crafty resets the canvas transform all by itself when the scene changes or the browser is resized. I need to apply a scale factor that overrides everything else for the lifetime of the game.
### One solution

The existing `Crafty.viewport.scale` method applies the new scale as a multiplier against the old scale. This pull request adds a `Crafty.viewport.absoluteScale` method that sets the scale absolutely, regardless of the old scale. Calling `scale(0.5)` twice results in a scale of 0.25. Calling `absoluteScale(0.5)` twice results in a scale of 0.5. This is useful when scaling the stage in response to browser window resize events. 

Unfortunately, this doesn't fix Crafty's incompatibility between viewport scaling and panning in a mixed Canvas/DOM setting (see [fiddle](http://jsfiddle.net/JUxyE/3/)). It looks like setting `Crafty.viewport.x = 5` (or panning) moves DOM elements by 5 screen pixels and canvas elements by 5 scaled pixels, causing alignment problems. The viewport system may be due for rethinking.
### Sample resizing code

Here's some sample code for a Crafty app that can scale the viewport in response to the browser window size ([fiddle here](http://jsfiddle.net/JUxyE/4/)):

``` javascript
var Screen = { w: 800, h: 480 };

Crafty.init();
Crafty.canvas.init();

var scaler = function() {
  var current = { w: Crafty.viewport.width, 
                  h: Crafty.viewport.height };

  // This will reset the scale when it adjusts the canvas element
  // dimensions, so do it right before setting the scale we want.
  Crafty.viewport.reload();

  if ( current.w > current.h ) {
    // Landscape, scale to fit height
    Crafty.viewport.absoluteScale( current.h / Screen.h );
  } else {
    // Portrait, scale to fit width
    Crafty.viewport.absoluteScale( current.w / Screen.w )
  }
};

// Avoid drawing entities before we're ready to scale.
Crafty.removeEvent( Crafty, window, 'resize', Crafty.viewport.reload );

// Adjust scale to fit inside resized window.
Crafty.addEvent( Crafty, window, 'resize', scaler );

// Run the scaler after scene changes, since they reset the scale.
Crafty.bind( 'SceneChange', scaler );
```
### Discussion

Thoughts? Is there a better way to go about this? :dancer:

starwed commented:
Yeah, some of the viewport code is a mess.    (Zoom seems to be completely broken!)

Why do you manually clear the screen in `absoluteScale()`?  Calling `Crafty.viewport.scale` within the function should also trigger a complete redraw.  If that doesn't work there's a bug lurking somewhere.

One problem with using this in response to resize events is that it won't be compatible for zooming/scaling for visual effects. That probably doesn't matter for your particular case, but I wonder if there's a better way to provide that functionality.

mildmojo commented:
`Why do you manually clear the screen in absoluteScale()?`

I use a full-screen stage (no dimensions passed to `init`). When this code scaled the stage down, bits that were drawn in the area no longer covered by the stage weren't being cleared. So I'd have a smaller, properly-rendered stage and a non-updating giant copy surrounding it. I didn't dig into the Crafty code to figure out exactly why it was happening, but I figured it might be because the dirty rects were scaled down and no longer covered their old locations, so that part of the image wasn't erased. The clear fixed it. I can probably probe deeper here.

Hmm, you're right about this being incompatible with regular zoom/scale. Maybe we can break the absolute scale apart from regular zoom/scale and let the latter figure the absolute scale into their calculations? Basically, let the absolute scale redefine 1.0 when resetting the canvas?

There are also some animation artifacts when scaling down, which may be related to dirty rects and fractional pixels. I haven't looked into that yet, either.

starwed commented:
When you call scale, it triggers a complete redraw; dirty rectangles are only used for partial redraws.  If there are artifacts left over, that's a bug in the redraw code.  (I have a couple of guesses as to why that might be.)

I think a complete revamp of how the viewport code works is in order.  It would be much more flexible (and not very difficulat) to decouple the implementation of viewport parameters (panning and scale) from the functions that set or animate those parameters.  For one thing, that'll mean we can deal with changing the viewport transformations only during a draw event, rather than (potentially) several times a frame.  It should also make it easier to find and fix the various viewport related bugs.  

So I'll hold off on doing anything with this PR just yet.  It's a good thing to bring up, though!

mildmojo commented:
I could've written that last bit more clearly. By "animation artifacts when scaling down", I meant that when I tween a sprite from X1 to X2 on a stage that's been scaled down, it leaves a trail around its edges. That's the bit I thought might be related to fractional pixels. It's different from the other problem of original-scale graphics left over outside the now-reduced canvas.

I'm totally in favor of seeing viewport code with way fewer surprises. =) Part of the reason for opening this PR was to see where things lie. We can close it if you like.

I've been looking at Phaser's [test suite](http://gametest.mobi/phaser/), and their [camera scale demo](http://gametest.mobi/phaser/index.php?f=camera%20scale.js&d=cameras) is really slick. If only they were entity/component and had mobile & text rendering figured out. =)

mildmojo committed:
Adds Crafty.viewport.absoluteScale().

The existing scale() method applies the new scale as a multiplier
against the old scale. This adds an absoluteScale() method that sets
the scale absolutely, regardless of the old scale. Calling .scale(0.5)
twice results in a scale of 0.25. Calling .absoluteScale(0.5) twice
results in a scale of 0.5. This is useful when scaling the stage in
response to browser window resize events (see example code in comment
header).

mpetrovic commented:
If everyone really hates it that much, I can take another crack at it. I
wrote the camera commands in the first place. If they suck, it's my fault.

On Mon, May 13, 2013 at 4:01 PM, Tim notifications@github.com wrote:

> I could've written that last bit more clearly. By "animation artifacts
> when scaling down", I meant that when I tween a sprite from X1 to X2 on a
> stage that's been scaled down, it leaves a trail around its edges. That's
> the bit I thought might be related to fractional pixels. It's a different
> problem from original-scale graphics left over outside the now-reduced
> canvas.
> 
> I'm totally in favor of seeing viewport code with way fewer surprises. =)
> Part of the reason for opening this PR was to see where things lie. We can
> close it if you like.
> 
> I've been looking at Phaser's test suite http://gametest.mobi/phaser/,
> and their camera scale demohttp://gametest.mobi/phaser/index.php?f=camera%20scale.js&d=camerasis really slick. If only they were entity/component and had mobile & text
> rendering figured out. =)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/craftyjs/Crafty/pull/463#issuecomment-17836622
> .

mildmojo commented:
Sorry about the dirty commit, too; just realized my text editor introduced a lot of noise when it stripped trailing whitespace. Fixed, if it matters.

mildmojo commented:
@mpetrovic No worries! I've typically done all my entity positioning by hand with a ton of individual resize callbacks. This was a result of diving into the viewport stuff after the [last Ludum Dare](http://www.ludumdare.com/compo/ludum-dare-26/?action=preview&uid=6748) to see if it could help me out.

Sometime in the future I'll try to make up a few fiddles of the ways the viewport behavior surprises me. It sounds like maybe the viewport features were intended for effects, and I'm trying to use them for utility, like UI alignment.

mpetrovic commented:
They were designed to mimic camera movements. Scale is actually closer to
the 'dolly' camera movement.

On Mon, May 13, 2013 at 4:13 PM, Tim notifications@github.com wrote:

> @mpetrovic https://github.com/mpetrovic No worries! I've typically done
> all my entity positioning by hand with a ton of individual resize
> callbacks. This was a result of diving into the viewport stuff after the last
> Ludum Darehttp://www.ludumdare.com/compo/ludum-dare-26/?action=preview&uid=6748to see if it could help me out.
> 
> Sometime in the future I'll try to make up a few fiddles of the ways the
> viewport behavior surprises me. It sounds like maybe the viewport features
> were intended for effects, and I'm trying to use them for utility, like UI
> alignment.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/craftyjs/Crafty/pull/463#issuecomment-17837363
> .

starwed commented:
@mpetrovic  Had you any thoughts on how to refactor this -- or done any work on it?  The current problem with the viewport isn't really how you call the methods, but bugs in how the backend works. 

My gut feeling is that we should do the following:
- Have the camera movement methods just change some variables representing the position and zoom level of the camera.  Oh, and probably trigger some "ViewportChange" event or the like.
- Have each render method then apply the correct transformations before rendering the scene.

That means the viewport code doesn't need to know about how the render methods work, and means that the transformations aren't applied more often than necessary.  It would let us easily have, say, a background layer that panned at a different rate than the foreground.  It also means that the logic for animating camera motion will be a lot easier to follow and debug.  :)

mpetrovic commented:
I've already done all of that. It's in the Layers branch in my fork, and
it's been there for months while I've been busy. It still needs some
hammering on and I think a few components need converting away from
EnterFrame, which I renamed to Tick.

On Sun, Jun 9, 2013 at 3:55 PM, starwed notifications@github.com wrote:

> @mpetrovic https://github.com/mpetrovic Had you any thoughts on how to
> refactor this -- or done any work on it? The current problem with the
> viewport isn't how you call the methods, but bugs in how the backend works.
> 
> My gut feeling is that we should do the following:
> - Have the camera movement methods just change some variables
>   representing the position and zoom level of the camera. Oh, and probably
>   trigger some "ViewportChange" event or the like.
> - Have each render method then apply the correct transformations
>   before rendering the scene.
> 
> That means the viewport code doesn't need to know about how the render
> methods work, and means that the transformations aren't applied more often
> than necessary. It would let us easily have, say, a background layer that
> panned at a different rate than the foreground. It also means that the
> logic for animating camera motion will be a lot easier to follow and debug.
> :)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/craftyjs/Crafty/pull/463#issuecomment-19172332
> .

starwed commented:
This PR is pretty out-of-date, and the viewport system has undergone some (if not all) of the changes/enhancements discussed in thread, so closing.
